tokio/src/io/util/copy.rs

```diff
@@ -59,19 +99,14 @@ impl CopyBuffer {
                     }
                 }
 
-                let n = buf.filled().len();
-                if n == 0 {
+                if self.cap == 0 {
                     self.read_done = true;
```

Comment:
You should move this check into the `poll_fill_buf` method. Currently we do not set `read_done` if the read happens inside `poll_write_buf`.